### DISCUSSION TITLE
Blog Teaser and Wikilinks not working

### DISCUSSION THREAD

**Discussion by AgedLace** [mkdocs-publisher/mkdocs-publisher#30] [General]
The Blog index shows the full posts of all posts.

I still cannot get the wikilinks to work.  I have double-checked  my obsidian settings - it is set to relative path to file.  When I do "mkdocs serve" I get the error message that all links are absolute links.  

When I click on the first link at the bottom of any post - it takes me to the full page of blog-post3 (last one). Clicking on the second link, or the post titles, I get a 404 page.

**Comment by mkusz** [comment]
Can you share your tests over some repo?

**Comment by AgedLace** [comment]
2nd repo is [here](https://github.com/AgedLace/MkDocs-Publisher-Take2)
[20230907_212208_mkdocs_build.log](https://github.com/mkusz/mkdocs-publisher/files/12569524/20230907_212208_mkdocs_build.log)
[20230907_212208_mkdocs_debug.zip](https://github.com/mkusz/mkdocs-publisher/files/12569525/20230907_212208_mkdocs_debug.zip)

**Reply by mkusz** [reply]
Thx. Will take a look later on that, but I have found a massive wiki link support error in my code. Fix in progress (it will also introduce some more functionality taken from Obsidian like image size manipulation).

**Reply by mkusz** [reply]
Hi, didn't have time to check yours repo,  but on [this branch](https://github.com/mkusz/mkdocs-publisher/tree/v1.1.1) you can find a new version of the plugin. The entire engine responsible for wiki links support has been rewritten and now should work correctly (even inside blog posts). If you will find time to test it, I will be highly appriciated.
PS. There are some unit tests added for some files, but it's just abeginningg of them and despite the tool showing like 27% coverage, in fact it's more like 5-7%, so it's a long jurney infront of me (and a lot of internal code refactor). I will finish this release next week and for next few weeks I will need to do some other stuff.

**Comment by AgedLace** [comment]
Okay - I'll work on this sometime this upcoming weekend and give you whatever feedback I can.  I totally understand needing to do other stuff.  

I will try to help you however / whenever I am able.

**Reply by AgedLace** [reply]
I have run `pip install git https://github.com/mkusz/mkdocs-publisher.git@v1.1.1` in my virtual environment and get the following error(s).

---

Collecting git+https://github.com/mkusz/mkdocs-publisher.git@v1.1.1
  Cloning https://github.com/mkusz/mkdocs-publisher.git (to revision v1.1.1) to /tmp/pip-req-build-el263rc9
  Running command git clone --filter=blob:none --quiet https://github.com/mkusz/mkdocs-publisher.git /tmp/pip-req-build-el263rc9
  Running command git checkout -b v1.1.1 --track origin/v1.1.1
  Switched to a new branch 'v1.1.1'
  branch 'v1.1.1' set up to track 'origin/v1.1.1'.
  Resolved https://github.com/mkusz/mkdocs-publisher.git to commit 18450e2e64ef6825d58d45335291971851f90adf
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-Take2/.venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-Take2/.venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-Take2/.venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/api.py", line 42, in prepare_metadata_for_build_wheel
          builder = WheelBuilder(poetry)
                    ^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/wheel.py", line 61, in __init__
          super().__init__(poetry, executable=executable)
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/builder.py", line 85, in __init__
          self._module = Module(
                         ^^^^^^^
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/utils/module.py", line 79, in __init__
          PackageInclude(
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/utils/package_include.py", line 29, in __init__
          self.check_elements()
        File "/tmp/pip-build-env-ddok_1hl/overlay/lib/python3.11/site-packages/poetry/core/masonry/utils/package_include.py", line 77, in check_elements
          raise ValueError(f"{root.name} is not a package.")
      ValueError: _cli is not a package.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

**Reply by mkusz** [reply]
Take a look at current branch. There should be fix for this problem and some other stuff. Just be aware of 2 things:
- mkdocs-material is now udpated to the latest one
- because of the above, I had to change one of the meta key name related to document publication and from now one instead of `status` there will a key named `publish` (the same name like in Obsidian) with possibility to use `true` (works the same as `published`) and `false` (works the same as `draft`).
There were also some small other updates but mostly some code internals.

**Comment by AgedLace** [comment]
I will get to this later in the week after taking care of other matters.  Thank you for your time and effort!

**Comment by AgedLace** [comment]
Tried running from this repo version with errors.  Build and Debug files attached ... also attaching my mkdocs.yml (renamed to .log so I can upload it here)  for your reference.  Seems there may be some issues there?  I changed all the "status" keys to "publish".  

 Also, once we iron out these issues, need to update documentation, especially with respect to the new publish key for status.  

[20230929_155730_mkdocs_build.log](https://github.com/mkusz/mkdocs-publisher/files/12768629/20230929_155730_mkdocs_build.log)
[20230929_155730_mkdocs_debug.zip](https://github.com/mkusz/mkdocs-publisher/
[mkdocs.yml.log](https://github.com/mkusz/mkdocs-publisher/files/12768657/mkdocs.yml.log)
files/12768631/20230929_155730_mkdocs_debug.zip)

**Reply by mkusz** [reply]
@AgedLace what kind of error do you have? I don't see any ERROR or WARNING message in the log. As for `mkdocs.yml` you don't have to put all configuration options from documentation into that file for each plugin. For example, if you want to use `pub-obsidian` you should only do:
```
plugins:
  - pub-obsidian
  ```
  and all default settings will be automatically applied. You have to use those settings, only when you want to change them according to your needs.
PS. Maybe you will again share a whole test repo with me.
PS2. Updated documentation will be part of the new release.

**Comment by AgedLace** [comment]
Here is the error which occurs after executing "mkdocs serve" in my virtual environment. 

---

(.venv) [deb@deb-desktop MkDocs-Publisher-4]$ mkdocs serve
WARNING -  Config value 'plugins': Plugin 'pub-meta' option 'slug': Sub-option 'enable': Unrecognised configuration name: enable
WARNING -  Config value 'plugins': Plugin 'pub-meta' option 'status': Unrecognised configuration name: status
WARNING -  Config value 'plugins': Plugin 'pub-obsidian' option 'links': Sub-option 'img_lazy': Unrecognised configuration name: img_lazy
11:57:30.375 [INFO ] Building documentation... [mkdocs.commands.serve]
11:57:30.375 [INFO ] Reading meta data from "README.md" files [mkdocs.plugins.publisher.meta.plugin]
11:57:30.380 [INFO ] Draft directories: ['blog', '.obsidian', '_templates'] [mkdocs.plugins.publisher.meta.plugin]
11:57:30.381 [INFO ] Hidden directories: [] [mkdocs.plugins.publisher.meta.plugin]
11:57:30.381 [INFO ] Draft files: [] [mkdocs.plugins.publisher.meta.plugin]
11:57:30.381 [INFO ] Hidden files: [] [mkdocs.plugins.publisher.meta.plugin]
11:57:30.417 [INFO ] Parsing blog posts from 'blog' directory [mkdocs.plugins.publisher.blog.parsers]
11:57:30.422 [INFO ] Creating blog posts index files [mkdocs.plugins.publisher.blog.creators]
11:57:30.483 [INFO ] Reorder blog posts from newest to oldest [mkdocs.plugins.publisher.blog.modifiers]
11:57:30.484 [INFO ] Cleaning site directory [mkdocs.commands.build]
11:57:30.606 [INFO ] Platform: Linux-6.1.53-1-MANJARO-x86_64-with-glibc2.38 [mkdocs.plugins.publisher.debug.plugin]
11:57:30.606 [INFO ] Python version: 3.11.5 (using virtual environment: true) [mkdocs.plugins.publisher.debug.plugin]
11:57:30.607 [INFO ] Build log file: /home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/20230929_155730_mkdocs_build.log [mkdocs.plugins.publisher.debug.plugin]
11:57:31.66. [INFO ] Debugger ZIP file: 20230929_155730_mkdocs_debug.zip [mkdocs.plugins.publisher.debug.plugin]
Traceback (most recent call last):
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/bin/mkdocs", line 10, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/__main__.py", line 270, in serve_command
    serve.serve(**kwargs)
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/commands/serve.py", line 86, in serve
    builder(config)
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/commands/serve.py", line 67, in builder
    build(config, live_server=None if is_clean else server, dirty=is_dirty)
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/commands/build.py", line 304, in build
    files = config.plugins.on_files(files, config=config)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/plugins.py", line 533, in on_files
    return self.run_event('files', files, config=config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs/plugins.py", line 507, in run_event
    result = method(item, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/deb/Work Directory/TestingGrounds/MkDocs-Publisher-4/.venv/lib/python3.11/site-packages/mkdocs_publisher/meta/plugin.py", line 328, in on_files
    file.url = quote(f"{'/'.join(url_parts)}/")
                        ^^^^^^^^^^^^^^^^^^^
TypeError: sequence item 1: expected str instance, bool found

---

I will create another repo for you to look at.  I didn't on this one because it wouldn't even render locally.  Will edit this post with the repo link once completed.

[Repo Link](https://github.com/AgedLace/mkdocs-publisher)

THANK YOU!!!

PS - I saw that material for mkdocs submitted a PR to the mkdocs/catalog github repo asking them to turn off the offline plugin in online contrexts as it was causing issues.  Have no idea if this has anything to do with the above error.  Just wanted to make you aware of it.

**Reply by mkusz** [reply]
Hi @AgedLace,
I have checked repo provided by you and have some conclusions:
- I have fixed debugger plugin, so now log file will catch all warnings related to config values
- sufficient plugin configuration in `mkdocs.yml` should look like this:
```yaml
plugins:
  - pub-debugger
  - search
  - pub-obsidian
  - pub-blog
  - pub-social
  - pub-meta
  - pub-minifier
```
- you should enable this setting: `use_directory_urls: true` since this setting is crucial to some of my plugins (need to test it more, if it's even possible to set it to `false`)
- according to [MkDocs documentation](https://www.mkdocs.org/getting-started/) `site_url` setting in `mkdocs.yml` is mandatory, you have to set it up (if you don't know your final url, provide some fake value like `site_url: https://fake.url/`).

Material for mkdocs offline plugin is not a problem here. Please read more carefully the beggining of the log, especially the parts where you have warnings related to config errors like those provided by you:
```
WARNING - Config value 'plugins': Plugin 'pub-meta' option 'slug': Sub-option 'enable': Unrecognised configuration name: enable
WARNING - Config value 'plugins': Plugin 'pub-meta' option 'status': Unrecognised configuration name: status
WARNING - Config value 'plugins': Plugin 'pub-obsidian' option 'links': Sub-option 'img_lazy': Unrecognised configuration name: img_lazy
```

**Comment by AgedLace** [comment]
Thank you for your patience in pointing out my errors!  I am not too techy and am still learning.

I had corrected all of the warnings that I understood; the ones that remained, I did not know how to fix.  

Have fixed mkdocs.yml per your instructions.

tested the following in the blog posts ...

- Backlinks work!
- External markdown links work
- Post Heading Links work
- Wikilinks do not work
- the post excerpt is still not working - entire blog posts are rendered in the blog index and archives
- Tags and Categories work as expected

Attaching build and debug logs.  [Repo address remains the same](https://github.com/AgedLace/mkdocs-publisher). 

[20231002_124103_mkdocs_build.log](https://github.com/mkusz/mkdocs-publisher/files/12782725/20231002_124103_mkdocs_build.log)
[20231002_124103_mkdocs_debug.zip](https://github.com/mkusz/mkdocs-publisher/files/12782726/20231002_124103_mkdocs_debug.zip)

**Reply by mkusz** [reply]
It seems like you have found the bug that was quite obvious, but I just didn't add a proper test case for it. I have assumed that all wiki links look like this `[[link_part|link text]]`, but in fact there is much more cases for it and one of them is just a `[[link_part_only]]`. There is a fix on the branch, so you can test it. It should work ;)
PS. I have added additional test cases to cover it, so it should be broken in the future, but if you will find anything new, please let me know.
PS2. Thanks for all your feedback. It's very helpfull.

**Comment by AgedLace** [comment]
Yes, the links are working exactly as you stated.  the [[link_part]] by itself does not work. The [[link_part|link text]] does work.  I have pushed those changes to the repo for your convenience.  

When I have more time,  I'll test callouts next as I use those pretty extensively.  I'll also work on creating a more thorough test file that incorporates as many features of both markdown and obsidian syntax so you have more thorough data to work with.

I am grateful that I am able to help you!  Thank you!!

**Reply by mkusz** [reply]
Now the `[[only_link]]` should also work. Reinstall library from the branch and test it.

**Comment by AgedLace** [comment]
Yes, now working with [[only_link]].  

Will now begin work on callouts, and a more thorough test document.

**Comment by AgedLace** [comment]
Have tested both MkDocs admonitions and Obsidian Callouts ... including foldable and nestable callouts.  They all work perfectly locally.  

QUESTION - do you only need copies of the build and debug logs when there are problems? Or do you want them each time?

Will work on a more complete markdown / obsidian syntax document tomorrow.

**Reply by mkusz** [reply]
They are needed only when something goes wrong. This way I can see if there is any error message inside the log file and if not, while fixing the problem, I can add more log messages in the code for potential future problem-solving.

**Comment by mkusz** [comment]
@AgedLace Closing this discussion since it's solved. If you will find any other problem, please open a new discussion or an issue.
